compiler/src/dotty/tools/dotc/transform/init/Errors.scala

```diff
@@ -0,0 +1,116 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import ast.tpd._
+
+import core._
+import Decorators._
+import Types._, Symbols._, Contexts._
+import util.NoSourcePosition
+
+import Effects._, Potentials._
+
+object Errors {
+  type Errors = Set[Error]
+  val empty: Errors = Set.empty
+
+  def show(errs: Errors)(implicit ctx: Context): String =
+    errs.map(_.show).mkString(", ")
+
+  sealed trait Error {
+    def source: Tree
+    def trace: Vector[Tree]
+    def show(implicit ctx: Context): String
+
+    def report(implicit ctx: Context): Unit =
+      ctx.warning(show + stacktrace, source.sourcePos)
+
+    def toErrors: Errors = Set(this)
+
+    def stacktrace(implicit ctx: Context): String = if (trace.isEmpty) "" else " Calling trace:\n" + {
+      var indentCount = 0
+      var last = ""
+      val sb = new StringBuilder
+      trace.foreach { tree =>
+        indentCount += 1
+        val pos = tree.sourcePos
+        val line = "[ " + pos.source.file.name + ":" + (pos.line + 1) + " ]"
+        if (last != line)
+          sb.append(
+            if (pos.source.exists)
+              i"${ " " * indentCount }-> ${pos.lineContent.trim}\t$line\n"
+            else
+              i"${tree.show}\n"
+          )
+        last = line
+      }
+      sb.toString
+    }
+
+    /** Flatten UnsafePromotion errors
+     */
+    def flatten: Errors = this match {
+      case unsafe: UnsafePromotion => unsafe.errors.flatMap(_.flatten)
+      case _ => Set(this)
+    }
+  }
+
+  /** Access non-initialized field */
+  case class AccessNonInit(field: Symbol, trace: Vector[Tree]) extends Error {
+    def source: Tree = trace.last
+    def show(implicit ctx: Context): String =
+      "Access non-initialized field " + field.show + "."
+
+    override def report(implicit ctx: Context): Unit = ctx.error(show + stacktrace, field.sourcePos)
+  }
+
+  /** Promote `this` under initialization to fully-initialized */
+  case class PromoteThis(pot: ThisRef, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String = "Promote `this` to be initialized while it is not."
```

Comment:
I think the wording "to be initialized" is unclear here and in the subsequent errors. Is it "promote this to (be initialized)" (as in "be initialized" is an initialization status) or is it "promote this (to be initialized)" (as in "promote this in order for it to become initialized")?

The scaladoc comments for `PromoteThis`, `PromoteWard` and `PromoteCold` have a better wording IMO.

Maybe being more verbose here would help: "Change `this` object's initialization status to 'fully initialized' while the object doesn't meet the criteria to have this status" (which criteria?).